src/charm.py

```diff
@@ -180,21 +188,23 @@ def _cli_args(self) -> str:
         """
         return "-config.file=/etc/agent/agent.yaml -prometheus.wal-directory=/tmp/agent/data"
 
-    def _config_file(self) -> str:
-        """Put all the config sections together and return the main config file.
+    def _config_file(self, event: EventBase) -> str:
+        """Generates config file str based on the event received.
 
         Returns:
-            The config file contents
+            A yaml string with grafana agent config
         """
         config = {}
         if server_config := self._server_config():
             config["server"] = server_config
+
         if integrations_config := self._integrations_config():
             config["integrations"] = integrations_config
+
         if prometheus_config := self._prometheus_config():
             config["prometheus"] = prometheus_config
-        if loki_config := self._loki_config():
-            config["loki"] = loki_config
+
+        self._loki_config(config, event)
         return yaml.dump(config)
```

Comment:
General comment: from first read it is difficult for me to follow what is going on (how a config is constructed), but it may be due to the complexity involved.

Reply:
Yes, you are right, it's not clear. I'll refactor it.